### ISSUE TITLE
[integers_patch] type 'Null' is not a subtype of type 'int' in int.tryParse

### ISSUE THREAD

**Issue by Coronon** [dart-lang/sdk#46110]
Trying to parse strings to ints can result in a `type 'Null' is not a subtype of type 'int'` RuntimeError.

I am running: ```Dart SDK version: 2.13.0 (stable) (Wed May 12 12:45:49 2021 +0200) on "windows_x64"```

Example code:
```dart
int.tryParse("10A");
```

This happens because of the following chain (Everything in [integers_patch.dart](https://github.com/dart-lang/sdk/blob/master/sdk/lib/_internal/vm/lib/integers_patch.dart)):
```dart
static int? tryParse(String source, ...) {
  // ...
  try {
      return _parse(unsafeCast<_StringBase>(source), radix, _kNull); // <- the _kNull is the problem here
    } catch (e) {
      return null;
    }
}

// Notice that this can not return 'null'
static int _parse(_StringBase source, int? radix, onError) {
  //...
  return _parseRadix(source, radix, start, end, sign, false, onError); // <- _kNull is 'onError'
}

// Notice that this can not return 'null'
static int _parseRadix(String source, int radix, int start, int end, int sign,
      bool allowU64, onError) {
  // ...
  return _throwFormatException(onError, source, start, radix, null); // <- The onError
  // ...
}

// Notice that this can not return 'null'
static int _throwFormatException(onError, source, index, radix, message) {
  if (onError != null) return onError(source); // <- This returns the result of calling _kNull which is null -> but _throwFormatException´s return type is non-nullable.
// **ERROR**
}
```

This error is catched by the `catch` in `tryParse`, but it is also reported every time this code runs when enabeling "all exceptions" in the dart debugger! I am astouned noeone else is experiencing this problem. It doesnt break my app, but it is extremly annoying when trying to debug the codebase.

**Comment by lrhn** [comment]
I believe this has already been fixed. It does not fail on the 2.13.0 SDK.

**Comment by Coronon** [comment]
> 
> 
> I believe this has already been fixed. It does not fail on the 2.13.0 SDK.

As said, this does happen in SDK 2.13.0, which is what I am running.

**Comment by lrhn** [comment]
That's interesting. What I get is:
```
C:\dart\> dart --version
Dart SDK version: 2.13.0 (stable) (Wed May 12 12:45:49 2021 +0200) on "windows_x64"

C:\dart\> type intpar.dart
// @dart = 2.12
main() {
  int? v = int.tryParse("10A");
  print(v);
}

C:\dart\> dart intpar.dart
null
```

The code has changed on tip-of-tree, so the linked code does not exist any more. I can see that it wasn't in 2.13.1, but will be in 2.14 (introduced on 2.14 track in https://github.com/dart-lang/sdk/blob/2.14.0-61.0.dev/sdk/lib/_internal/vm/lib/integers_patch.dart).

So, it is fixed, but not in 2.13.

**Comment by a-siva** [comment]
@lrhn is this something that needs to be cherry picked into 2.13 ?
